ibis/expr/operations.py

```diff
@@ -75,30 +76,45 @@ def flat_args(self):
             else:
                 yield arg
 
+    def __hash__(self):
+        if not hasattr(self, '_hash'):
+            self._hash = hash(
+                tuple(
+                    element.op() if isinstance(element, ir.Expr) else element
+                    for element in self.flat_args()
+                )
+            )
+        return self._hash
+
+    def __eq__(self, other):
+        return self.equals(other)
+
     def equals(self, other, cache=None):
         if cache is None:
             cache = {}
 
-        if (self, other) in cache:
-            return cache[(self, other)]
+        try:
+            return cache[self, other]
+        except KeyError:
+            pass
 
         if self is other:
-            cache[(self, other)] = True
+            cache[self, other] = True
             return True
 
         if type(self) != type(other):
-            cache[(self, other)] = False
+            cache[self, other] = False
             return False
 
         if len(self.args) != len(other.args):
-            cache[(self, other)] = False
+            cache[self, other] = False
             return False
 
         for left, right in zip(self.args, other.args):
```

Comment:
You could use simply:

```python
all_equal(self.args, other.args, cache=cache)
```

But the length of the arguments needs to be checked here as well, because [map](https://github.com/ibis-project/ibis/blob/42cbdc0783b5cd9f5f9b1a6a7c763244b3b6bd7f/ibis/expr/operations.py#L159) zips its arguments.

Reply:
The equality PR will address this length issue. I will refactor this and the code near your other comment to use `all_equal` once that is PR is merged.